sdk/src/azure/platform/az_mosquitto.c

```diff
@@ -0,0 +1,278 @@
+      me,
+      &(az_mqtt_disconnect_data){ .tls_authentication_error = false,
+                                  .disconnect_requested = (rc == 0) });
+
+  if (az_result_failed(ret))
+  {
+    _az_mosquitto_critical_error();
+  }
+}
+
+/* Callback called when the client knows to the best of its abilities that a
+ * PUBLISH has been successfully sent. For QoS 0 this means the message has
+ * been completely written to the operating system. For QoS 1 this means we
+ * have received a PUBACK from the broker. For QoS 2 this means we have
+ * received a PUBCOMP from the broker. */
+static void _az_mosquitto_on_publish(struct mosquitto* mosq, void* obj, int mid)
+{
+  (void)mosq;
+  az_mqtt* me = (az_mqtt*)obj;
+
+  az_result ret = az_mqtt_inbound_puback(me, &(az_mqtt_puback_data){ mid });
+
+  if (az_result_failed(ret))
+  {
+    _az_mosquitto_critical_error();
+  }
+}
+
+static void _az_mosquitto_on_subscribe(
+    struct mosquitto* mosq,
+    void* obj,
+    int mid,
+    int qos_count,
+    const int* granted_qos)
+{
+  (void)mosq;
+  (void)qos_count;
+  (void)granted_qos;
+
+  az_mqtt* me = (az_mqtt*)obj;
+
+  az_result ret = az_mqtt_inbound_suback(me, &(az_mqtt_suback_data){ mid });
+
+  if (az_result_failed(ret))
+  {
+    _az_mosquitto_critical_error();
+  }
+}
+
+static void _az_mosquitto_on_unsubscribe(struct mosquitto* mosq, void* obj, int mid)
+{
+  (void)mosq;
+  (void)obj;
+  (void)mid;
+
+  // Unsubscribe is not handled or implemented. We should never get here.
+  _az_mosquitto_critical_error();
+}
+
+static void _az_mosquitto_on_message(
+    struct mosquitto* mosq,
+    void* obj,
+    const struct mosquitto_message* message)
+{
+  (void)mosq;
+  az_mqtt* me = (az_mqtt*)obj;
+
+  az_result ret = az_mqtt_inbound_recv(
+      me,
+      &(az_mqtt_recv_data){ .qos = (int8_t)message->qos,
+                            .id = (int32_t)message->mid,
+                            .payload = az_span_create(message->payload, message->payloadlen),
+                            .topic = az_span_create_from_str(message->topic) });
+
+  if (az_result_failed(ret))
+  {
+    _az_mosquitto_critical_error();
+  }
+}
+
+static void _az_mosquitto_on_log(struct mosquitto* mosq, void* obj, int level, const char* str)
+{
+  (void)mosq;
+  (void)obj;
+  (void)level;
+  if (_az_LOG_SHOULD_WRITE(AZ_LOG_MQTT_STACK))
+  {
+    _az_LOG_WRITE(AZ_LOG_MQTT_STACK, az_span_create_from_str((char*)(uintptr_t)str));
+  }
+}
+
+AZ_NODISCARD az_mqtt_options az_mqtt_options_default()
+{
+  return (az_mqtt_options){ .certificate_authority_trusted_roots = NULL,
+                            .openssl_engine = NULL,
+                            .mosquitto_handle = NULL };
+}
+
+AZ_NODISCARD az_result az_mqtt_init(az_mqtt* mqtt, az_mqtt_options const* options)
+{
+  _az_PRECONDITION_NOT_NULL(mqtt);
+  mqtt->_internal.options = options == NULL ? az_mqtt_options_default() : *options;
+  mqtt->mosquitto_handle = mqtt->_internal.options.mosquitto_handle;
```

Comment:
Why do we need a copy of the mosquitto_handle in the obj? Wouldn't we always have this if we have access to the obj from mosquitto?

Reply:
We don't always have access: our public API can be called on any thread (application thread, timer callback thread, mosquitto callback thread, etc.)
How would `az_mqtt_outbound_connect` work for example?